sentence_transformers/losses/CachedMultipleNegativesSymmetricRankingLoss.py

```diff
@@ -0,0 +1,268 @@
+from __future__ import annotations
+
+from contextlib import nullcontext
+from functools import partial
+from typing import Any, Iterable, Iterator
+
+import torch
+import tqdm
+from torch import Tensor, nn
+
+from sentence_transformers import SentenceTransformer, util
+from sentence_transformers.losses.CachedMultipleNegativesRankingLoss import RandContext
+
+
+def _backward_hook(
+    grad_output: Tensor,
+    sentence_features: Iterable[dict[str, Tensor]],
+    loss_obj: CachedMultipleNegativesSymmetricRankingLoss,
+) -> None:
+    """A backward hook to backpropagate the cached gradients mini-batch by mini-batch."""
+    assert loss_obj.cache is not None
+    assert loss_obj.random_states is not None
+    with torch.enable_grad():
+        for sentence_feature, grad, random_states in zip(sentence_features, loss_obj.cache, loss_obj.random_states):
+            for (reps_mb, _), grad_mb in zip(
+                loss_obj.embed_minibatch_iter(
+                    sentence_feature=sentence_feature,
+                    with_grad=True,
+                    copy_random_state=False,
+                    random_states=random_states,
+                ),
+                grad,
+            ):
+                surrogate = torch.dot(reps_mb.flatten(), grad_mb.flatten()) * grad_output
+                surrogate.backward()
+
+
+class CachedMultipleNegativesSymmetricRankingLoss(nn.Module):
+    def __init__(
+        self,
+        model: SentenceTransformer,
+        scale: float = 20.0,
+        similarity_fct: callable[[Tensor, Tensor], Tensor] = util.cos_sim,
+        mini_batch_size: int = 32,
+        show_progress_bar: bool = False,
+    ) -> None:
+        """
+        Cached version of MultipleNegativesSymmetricRankingLoss with GradCache optimization.
+
+        This loss is an adaptation of MultipleNegativesRankingLoss and CachedMultipleNegativesRankingLoss.
```

Comment:
> This loss is an adaptation of MultipleNegativesRankingLoss and CachedMultipleNegativesRankingLoss.

Should this have been MultipleNegativesSymmetricRankingLoss and CachedMultipleNegativesRankingLoss?

Reply:
That's correct, thanks.